### PR TITLE
Mixpanel/identity management

### PR DESCRIPTION
**Pull request by icerove**
Cli Identity Management + Consent
Goal: 
Get account ID from users who consent so that we can
Get action + error history for partners / those asked for help --- a more efficient way to help with more context
Understand Cli user journey better for improvements
A unique ID for future full developer journey mapping connecting website, docs, examples, wallets etc.

Implementation Part 1: Consent function

- [x] Add “consent to expose ID to NEAR” in the current AskForConsent Function

Implementation Part 2: Account ID auto storing command
If the user opt in to our tracking, we will only implement tracking code to extract/store their account ID in the commands below: 

- [x] Login <Start + Success>
- [x] NEAR Create Account(wait for discussion to get master or new account)
- [x] Dev Deploy
- [x] Deploy
- [x] Shell Account Status Lookup Started (NEAR State)
- [x] NEAR repl

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -30,6 +30,7 @@ exports.clean = async function () {
 };
 
 exports.deploy = async function (options) {
+    await eventtracking.askForOptInAndAccountID(options);
```
Comment (vgrichina) [inline index.js:33]: this is not necessarily called by user, actually more likely to be used in some deploy script on CI, etc

File: test/test_account_operations.sh [hunk 2]
```diff
@@ -9,7 +9,7 @@ echo Create account
 ../bin/near create-account $testaccount
 
 echo Get account state
-RESULT=$(../bin/near state $testaccount -v | ../node_modules/.bin/strip-ansi)
+RESULT=$(yes | ../bin/near state $testaccount -v | ../node_modules/.bin/strip-ansi)
```
Comment (vgrichina) [inline test/test_account_operations.sh:12]: shouldn't require user interaction, i.e. no `yes |`

File: test/test_generate_key.sh [hunk 3]
```diff
@@ -5,7 +5,7 @@ KEY_FILE=~/.near-credentials/$NODE_ENV/generate-key-test.json
 rm -f "$KEY_FILE"
 echo "Testing generating-key: new key"
 
-RESULT=$(./bin/near generate-key generate-key-test --networkId $NODE_ENV -v)
+RESULT=$(yes |./bin/near generate-key generate-key-test --networkId $NODE_ENV -v)
```
Comment (vgrichina) [inline test/test_generate_key.sh:8]: shouldn't require user interaction, i.e. no `yes |`

File: utils/eventtracking.js [hunk 4]
```diff
@@ -109,7 +122,58 @@ const getEventTrackingConsent = async () => {
     }
 };
 
-const askForConsentIfNeeded = async (options) => {
+const getIdTrackingConsent = async () => {
```
Comment (vgrichina) [inline utils/eventtracking.js:125]: I think we should only ask people one question, i.e. it's enough to just add text in question above in `getEventTrackingConsent`

File: utils/eventtracking.js [hunk 5]
```diff
@@ -109,7 +122,58 @@ const getEventTrackingConsent = async () => {
...
+    const rl = readline.createInterface({
+        input: process.stdin,
+        output: process.stdout,
+    });
+    try {
+        for (let attempts = 0; attempts < 10; attempts++) {
+            const answer = await new Promise((resolve) => {
+                rl.question(
+                    chalk`We would like to help with your development journey with NEAR.` +
+                        chalk` We will ask you to share your account Id while using command. ` +
```
Comment (vgrichina) [inline utils/eventtracking.js:135]: nit: I think we usually spell it as `account ID`, e.g. https://docs.near.org/docs/concepts/account#account-id-rules

Comment (vgrichina) [reply utils/eventtracking.js:135]: I think we also need to add a text to explain to user that this data is already generally on chain. Something like:

`Note that your account ID and all associated on-chain transactions are already being recorded on public blockchain.`

Maybe @kcole16 @corwinharrell can help with text here. Basically we should try to avoid surprises.

File: utils/eventtracking.js [hunk 6]
```diff
@@ -109,7 +122,58 @@ const getEventTrackingConsent = async () => {
...
+
+const askForId = async (options) => {
+    const shellSettings = settings.getShellSettings();
+    if(shouldTrackID(shellSettings)){
+        const id = isGitPod() ? getGitPodUserHash() : shellSettings[TRACKING_SESSION_ID_KEY];
+        await Promise.all([
+            mixpanel.alias(options.accountId, id),
+            mixpanel.people.set(id, {account_id: options.accountId})
+        ]);
+    }else if(shouldNotTrackID(shellSettings)){
```
Comment (vgrichina) [inline utils/eventtracking.js:167]: nit, please format like in most other places:

```
} else if (shouldNotTrackID(shellSettings)) {
```

File: utils/eventtracking.js [hunk 7]
```diff
@@ -109,7 +122,58 @@ const getEventTrackingConsent = async () => {
...
+        return;
+    }
```
Comment (vgrichina) [inline utils/eventtracking.js:169]: nit:

```
} else {
```

File: commands/generate-key.js [hunk 8]
```diff
@@ -1,45 +1,50 @@
 const KeyPair = require('near-api-js').KeyPair;
```
Comment (vgrichina) [inline commands/generate-key.js:1]: changes in this file seem to be unrelated to the topic of PR, best to revert them

File: commands/dev-deploy.js [hunk 9]
```diff
@@ -36,7 +36,7 @@ module.exports = {
 };
 
 async function devDeploy(options) {
-    await eventtracking.askForConsentIfNeeded(options);
```
Comment (vgrichina) [inline commands/dev-deploy.js:39]: I'd prefer if we keep `askForConsentIfNeeded` name as it is open ended and allows to add more questions if needed. We don't want to rename it to `askForOptInAndAccountIDAndGeoLocation` next time we want to collect some other data.

**Review by vgrichina — CHANGES_REQUESTED** [review]
thanks, moving in right direction and good catch with `near repl`

there are few changes needed however

File: index.js [hunk 10]
```diff
@@ -175,6 +176,7 @@ exports.login = async function (options) {
 };
 
 exports.viewAccount = async function (options) {
+    await eventtracking.askForOptInAndAccountID(options);
```
Comment (vgrichina) [inline index.js:179]: this can be used in scripts, not necessarily user-interactive

File: utils/eventtracking.js (hunk 4 above)
Comment (icerove) [reply utils/eventtracking.js:125]: I do not think so, one question is to generate session id and another one is to track account id which is not the same. @vgrichina

File: utils/eventtracking.js (hunk 5 above)
Comment (icerove) [reply utils/eventtracking.js:135]: I added

File: index.js (hunk 10 above)
Comment (icerove) [reply index.js]: @vgrichina what do you mean using in scripts?

Comment (vgrichina) [reply index.js]: I mean that I can write scripts relying on `near-cli` (like e.g. we have scripts in projects created by create-near-app to use `near deploy`, etc).

E.g. I can try to get formatted amount from this call:

```
[I] ➜ near state account.testnet | grep formattedAmount
  formattedAmount: '328.96507017987727742332'
```

this kind of usage breaks if you have a prompt to user. 

We have few commands where user input is expected in any case (like `near login)` – so these can be used to ask user. Commands which don't necessary require user interaction shouldn't be interfered with. Somebody might use them in CI environment where there is no user to prompt. `near deploy` is quite likely to be used in CI for example.

Comment (vgrichina) [reply index.js]: @icerove like e.g. we can see this problem in action on near-cli CI:
https://travis-ci.com/github/near/near-cli/jobs/457657341

it basically hangs on `create-account` call

File: test/test_account_operations.sh (hunk 2 above)
Comment (vgrichina) [reply test/test_account_operations.sh]: this is not fixed, why it is resolved?

Comment (vgrichina) [reply test/test_account_operations.sh]: my bad actually, looks like it is fixed here. But tests don't pass on CI because we still ask for consent in commands which are used in scripts.

File: index.js (hunk 10 above)
Comment (icerove) [reply index.js]: the function `askForConsentIfNeeded` is asking questions to user instead of let user typing in some account ID or give consent from their side. If I understand correctly, writing scripts is not the case we want for asking process. @vgrichina
